### PR TITLE
Report workflow

### PR DESCRIPTION
**Pull request by msouff**

### CODE REVIEW

File: tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+from tethysext.atcore.controllers.utilities import get_plot_object_from_result, get_tabular_data_for_previous_steps
+
+from tethys_sdk.gizmos import DataTableView
+from collections import OrderedDict
+
+
+log = logging.getLogger(__name__)
+
+
+class ReportWorkflowResultsView(MapWorkflowView, WorkflowResultsView):
```
Comment (swainn) [inline tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py:24]: If you are creating your own Leaflet map for the report view, then You shouldn't need to inherit from MapWorkflowView here. Do you need this?

File: tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        height: 500px;
+        padding-left: 25px;
+        padding-top: 25px;
+      }
+      .tab-data {
+        display: block;
+      }
+    </style>
+  {% endif %}
+  <style>
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html:32]: This style block is long enough that it should really be moved out into a separate css file.

File: tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        {% for k, v in tab_data.items %}
+        <p><b>{{ k|title }}:</b> {{ v }}</p>
+        {% endfor %}
+        {% endfor %}
+      </div>
+    {% endif %}
+    <div contentEditable=true class="add-border" data-ph="General Note"></div>
+    <!-- Try to find a way to move the script (one line below) into the component. -->
+    <script>
+      var maps = [];
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html:175]: What are these variables used for?

File: tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+          </div>
+        {% endif %}
+      {% endfor %}
+    {% endfor %}
+  </div>
+  </form>
+{% endblock %}
+
+{% block scripts %}
+  {{ block.super }}
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html:227]: Remove any blocks that contain only `block.super` entries in them.

File: tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html [hunk 5]
```diff
@@ -0,0 +1,234 @@
+{% extends "atcore/resource_workflows/workflow_results_view.html" %}
+{% load tethys_gizmos static %}
+
+{% block app_content_wrapper_override %}
+  {{ block.super }}
+  <input type="button" value="Print Report" id="btnPrint" class="btn btn-info" style="position:absolute;right:20px;top:70px;"/>
+  <script type="text/javascript">
+    $("#btnPrint").on("click", function() {
+      let curURL = window.location.href;
+      history.replaceState(history.state, '', '/steem-report');
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html:10]: We cannot hard-code "steem" anywhere in atcore. Perhaps "print-report"?

File: tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+      window.print();
+      history.replaceState(history.state, '', curURL);
+    });
+  </script>
+{% endblock %}
+
+{% block content_dependent_styles %}
+  {{ block.super }}
+  {% if has_tabular_data %}
+    <style>
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html:20]: Again, if you are making your own maps, and not using the map view, then these styles are not needed (where we shrink the map to show the tabular data).

File: tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+  {% import_gizmo_dependency datatable_view %}
+  {% import_gizmo_dependency map_view %}
+{% endblock %}
+
+{% block app_content %}
+  {{ block.super }}
+
+  <form id="report_form">
+  <div id="report_workflow_data">
+    {% if has_tabular_data %}
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html:162]: Is anything in this if statement used/needed? Seems like this was copied over from the condor job view.

File: tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+{% endblock %}
+
+{% block scripts %}
+  {{ block.super }}
+{% endblock %}
+
+
+
+{% block after_app_content %}
+  {{ block.super }}
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html:233]: Another block that can be removed (232-234)

File: tethysext/atcore/public/resource_workflows/report_workflow_leaflet_map.js [hunk 9]
```diff
@@ -0,0 +1,27 @@
+// Create leaflet map.
+var map = new L.Map('map').setView([-41.59490508367679, 146.77734375000003], 7);
```
Comment (swainn) [inline tethysext/atcore/public/resource_workflows/report_workflow_leaflet_map.js:2]: Is this file needed?

File: tethysext/atcore/templates/atcore/resource_workflows/components/report_workflow_map_component.html [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+</div>
+<script>
+  wmsSources[{{ forloop.parentloop.counter }}] = new ol.source.ImageWMS({url: '{{ map_layer.options.url|safe }}',
+                                          params: {{ map_layer.options.params|safe }},
+                                          ratio: 1,
+                                          serverType: 'geoserver',
+                                        })
+  maps[{{ forloop.parentloop.counter }}] = new ol.Map({
+        interactions: ol.interaction.defaults({
+          //mouseWheelZoom:false,
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/resource_workflows/components/report_workflow_map_component.html:26]: Remove commented out lines.

File: tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+            request=request,
+            session=session,
+            context=context,
+            resource=resource,
+            current_step=current_step,
+            previous_step=previous_step,
+            next_step=next_step
+        )
+
+    def get_context(self, request, session, resource, context, model_db, workflow_id, step_id, result_id, *args,
```
Comment (swainn) [inline tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py:166]: Don't need this method if not inheriting from both MapWorkflowView and ResultWorkflowView (see previous comment).

File: tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+        """
+        # Turn off feature selection on model layers
+        map_view = context['map_view']
+        self.set_feature_selection(map_view=map_view, enabled=False)
+
+        # Can run workflows if not readonly
+        can_run_workflows = not self.is_read_only(request, current_step)
+
+        # get tabular data if any
+        tabular_data = get_tabular_data_for_previous_steps(
```
Comment (swainn) [inline tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py:54]: Is tabular_data needed for the report result view? Seems like it is here to satisfy a need in the template that isn't needed.

**Review by swainn — CHANGES_REQUESTED** [review]
Great first pass! It needs to be cleaned up a bit though. Please see comments.

File: tethysext/atcore/controllers/resource_workflows/map_workflows/spatial_condor_job_mwv.py [hunk 13]
```diff
@@ -73,23 +73,6 @@ def process_step_options(self, request, session, context, resource, current_step
             next_step=next_step
         )
 
-    def get_tabular_data_for_previous_steps(self, current_step):
```
[the remark this answers is not in the document]
Comment (msouff) [reply tethysext/atcore/controllers/resource_workflows/map_workflows/spatial_condor_job_mwv.py:76]: The thing is that ResourceWorkflow is not a parent for SpatialCondorJobMWV. adding it just so we can have that functions seems a little overkill, no? Take a look at what else was already in the utilities before you decide.

File: tethysext/atcore/controllers/resource_workflows/results_views/plot_workflow_results_view.py [hunk 14]
```diff
@@ -58,132 +50,131 @@ def get_context(self, request, session, resource, context, model_db, workflow_id
...
         result = self.get_result(request=request, result_id=result_id, session=session)
 
-        # Get datasets.
-        datasets = result.datasets
-
-        # Get plot object
-        plot_object = result.plot
-
+        plot_view = get_plot_object_from_result(result)
+        # # Get datasets.
```
[the remark this answers is not in the document]
Comment (msouff) [reply tethysext/atcore/controllers/resource_workflows/results_views/plot_workflow_results_view.py:54]: oops

File: tethysext/atcore/controllers/resource_workflows/map_workflows/spatial_condor_job_mwv.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (swainn) [reply tethysext/atcore/controllers/resource_workflows/map_workflows/spatial_condor_job_mwv.py:76]: You are right, it is not a parent of the view because it is a model. ResourceWorkflow is the workflow model, which you will have access to, no? Look at the other methods on ResourceWorkflow. Seems like a good fit and it will be available anywhere you have a workflow.

File: tethysext/atcore/public/gizmos/slide_sheet/slide_sheet.js [hunk 15]
```diff
@@ -26,11 +26,15 @@ var SLIDE_SHEET = (function() {
     *                    PRIVATE FUNCTION IMPLEMENTATIONS
     *************************************************************************/
     open = function(id) {
-        $('#' + id + '.slide-sheet').addClass('show');
+        if (id.length) {
```
[the remark this answers is not in the document]
Comment (htran84) [reply tethysext/atcore/public/gizmos/slide_sheet/slide_sheet.js:29]: For some reason, tethys sometimes tries to load the slide-sheet with no id given (empty string). This will cause the rest of the javascript to stop running. I put the if statement there to insure that the id must be specified to show the slidesheet.

File: tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html (hunk 3 above)
Comment (msouff) [reply tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html:175]: these are used to loop over results when there is more than one map, table ,etc.

File: tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html (hunk 7 above)
Comment (msouff) [reply tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html:162]: this is why we need the tabular data. This adds the title, general description and parameters to the top of the report.

Comment (swainn) [reply tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html:35]: Ah! Got you. Ok. That makes sense. Thanks for the clarification.

File: tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+                                                                                        into=OrderedDict)],
+                        searching=False,
+                        paging=False,
+                        info=False
+                    )
+                    ds.update({'data_table': data_table})
+                    ds.update({'data_description': result.description})
+                    results.append({'dataset': ds})
+            elif isinstance(result, PlotWorkflowResult):
+                results.append({'plot': [result.name, result.description, result.get_plot_object()]})
```
Comment (swainn) [inline tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py:82]: Use a dictionary instead of a list so that it is easier to follow when used in the template. For example:

```python
[result.name, result.description, result.get_plot_object()]
```
ends ups being used like this in the template:

```html
<h3>{{ v.0 }}</h3>
 {% if v.1 %}
     <h4>Description: {{ v.1 }}</h4>
 {% endif %}
<div id="resultPlotDiv{{ forloop.counter }}" style="width:800px;">
     {% gizmo v.2 %}
</div>
```

If you change it to a dictionary like:

```python
{'name': result.name, 'description': result.description, 'gizmo': result.get_plot_object()}
```

Then the template is easier to read:

```html
<h3>{{ v.name }}</h3>
 {% if v.description %}
     <h4>Description: {{ v.description }}</h4>
 {% endif %}
<div id="resultPlotDiv{{ forloop.counter }}" style="width:800px;">
     {% gizmo v.gizmo %}
</div>
```

File: tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py [hunk 17]
```diff
@@ -0,0 +1,218 @@
...
+            resource=resource,
+        )
+
+        # Get DatasetWorkflowResult
+        results = list()
+        for result in current_step.results:
+            if isinstance(result, DatasetWorkflowResult):
+                for ds in result.datasets:
+                    # Check if the export options is there
+                    data_table = DataTableView(
```
Comment (swainn) [inline tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py:70]: I'd split lines 70-79 into a separate method on this class and call it something like: build_dataset_result_dict or similar.

**Review by swainn — CHANGES_REQUESTED** [review]
Second round of comments in.

File: tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py (hunk 12 above)
Comment (swainn) [reply tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py]: Got it. Thanks for the clarification.

File: tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+                        paging=False,
+                        info=False
+                    )
+                    ds.update({'data_table': data_table})
+                    ds.update({'data_description': result.description})
+                    results.append({'dataset': ds})
+            elif isinstance(result, PlotWorkflowResult):
+                results.append({'plot': [result.name, result.description, result.get_plot_object()]})
+            elif isinstance(result, SpatialWorkflowResult):
+                params = ""
```
[the remark this answers is not in the document]
Comment (msouff) [reply tethysext/atcore/controllers/resource_workflows/results_views/report_workflow_results_view.py:84]: This one has changed a lot with the dynamic legend tasks. I propose we refine it in that PR instead. Else it will be double work.

File: tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+        {% for parameter, value in tab_data.items %}
+        <p><b>{{ parameter|title }}:</b> {{ value }}</p>
+        {% endfor %}
+        {% endfor %}
+      </div>
+    {% endif %}
+    <div contentEditable=true class="add-border" data-ph="General Note"></div>
+    <!-- Try to find a way to move the script (one line below) into the component. -->
+    <script>
+      var maps = [];
```
Comment (swainn) [inline tethysext/atcore/templates/atcore/resource_workflows/report_workflow_results_view.html:48]: Move these variables to the top of the report_workflow_results_view.js

**Review by swainn — COMMENTED** [review]
One last comment.

**Review by swainn — APPROVED** [review]
